### ISSUE TITLE
Is it possible to extract the functionality into a nuget package?

### ISSUE THREAD

**Issue by richorama** [OrleansContrib/Orleans.Indexing#1]
I was wondering how much of the indexing functionality is in the 'core' orleans code, and how much could be extracted to a packge. I would make this functionality easier to consume if it was an 'optional extra' on top of the core orleans codebase.

**Comment by creyke** [comment]
Agreed. Maybe the dependent abstractions could be PR-ed into the main Orleans code?

**Comment by Arshia001** [comment]
Was gonna ask for this just now.

**Comment by galvesribeiro** [comment]
I was going to look at it once I finish the new telemetry APIs. 

The big challenge now is to understand the code and make it injectable thru facets and not depend on a base class.

**Comment by philbe** [comment]
@richorama There's only a small amount of change to the Orleans runtime. Almost all of it runs at app level.

@mdashti, can you please summarize these changes to the runtime?

@galvesribeiro, yes, it should be made injectable thru facets. We developed the indexing prototype over a year ago, before the decision was made (last January) to refactor the transaction mechanism using injectable facets.

**Comment by galvesribeiro** [comment]
@ph1ll Yep, I'll look at it

**Comment by mdashti** [comment]
@richorama, thank you for bringing up this issue. As @ph1ll mentioned, we had this in our mind from the very beginning. These are three clean commits that should be included in the core:

- 72fd56b261bc542fe947c0436e5e915b4522c425
- d4df9507fafdb3cdb9e8cea876059b1ba68c5422
- e4c3479b4c9f8cb438a629611c36dab729504b39

The rest is completely separated into the OrleansIndexing sub-project.

**Comment by philbe** [comment]
FYI, my GitHub username is @philbe, not @ph1ll.  It looks like this got started via an autocomplete error by @galvesribeiro.

**Comment by galvesribeiro** [comment]
Ops, my bad @philbe  :)

**Comment by philbe** [comment]
@Arshia001 @richorama @creyke @galvesribeiro 
I have started working with a developer to port the indexing system to Orleans 2.0, to replace the base class by facets (i.e., dependency injection), and to obtain consistency between grains and indexes using transactions (in place of or in addition to the current workflow protocol).

**Comment by jamiemitchellconsultants** [comment]
@philbe 
If you are still progressing with Indexing, I can offer some time to help even if it is only writing tests, documentation, sample implementations

**Comment by philbe** [comment]
@jamiemitchellconsultants 
CC: @kirkolynyk, who is doing the port to Orleans 2.0
Thanks for the offer, but we're not far enough along for that -- still working on getting the code to run on 2.0. I'll report on the project's status next week.

**Comment by KSemenenko** [comment]
I'm very interested in this project, let me know if you need any help.

**Comment by KSemenenko** [comment]
@philbe any updates?

**Comment by philbe** [comment]
@TedHartMS has finished porting the implementation to Orleans V2. He's currently doing a final round of tests, after which we'll publish it as a new project on GitHub. (There were too many changes to Orleans to rebase the current Orleans.Indexing on Orleans V2.)  Sorry for the delay. It took longer than we expected because Orleans V2 doesn't expose all of the APIs that are needed by Orleans.Indexing, so we've been iterating with the Orleans team to agree on a minimal the number of changes that will eventually be needed to the Orleans runtime.

**Comment by KSemenenko** [comment]
@philbe any updates?

**Comment by philbe** [comment]
@KSemenenko We have a version that runs on Orleans V2. I expect we'll be publishing it early next week.

**Comment by philbe** [comment]
@KSemenenko @jamiemitchellconsultants @Arshia001 @richorama @creyke @galvesribeiro @mdashti @

@TedHartMS has published a new version of Orleans.Indexing that works on Orleans V2. It’s a new repo, OrleansV2.Fork.Indexing. Given the amount of churn in Orleans and Orleans.Indexing, it didn’t make sense to go through the effort of rebasing Orleans.Indexing and then merging in the changes.

OrleansV2.Fork.Indexing is a fork of Orleans V2 because it depends on some changes to the Orleans runtime. We think these changes are not controversial and will be accepted as soon as Sergey’s team has time. After they’re in Orleans, we’ll create another repo, OrleansV2.Indexing that uses the Orleans NuGet, and then delete OrleansV2.Fork.Indexing.

There’s still one missing feature to be added: DirectStorageManagedIndex, which uses Document DB (Cosmos DB). @TedHartMS needs to submit a pull request for Orleans.CosmosDB to get it working.
Next, we’ll be working on replacing the inheritance-based API (IIndexedGrain) by a facet-based interface, as was done for transactions, and then implement transactional indexing

**Comment by Arshia001** [comment]
Best news so far this week, thanks!

**Comment by KSemenenko** [comment]
Thank you @philbe  !! it's just amazing news!

**Comment by KSemenenko** [comment]
Is there any news? when will it be available officially? @philbe

**Comment by Arshia001** [comment]
My question as well. I've been using my own (naive) index system based on buckets (also implememted as grains), but I'd love an official solution.

**Comment by jamiemitchellconsultants** [comment]
@Arshia001 @KSemenenko  I have also been using my own naive index based on buckets implemented as grains... any appetite for sharing code ?

**Comment by Arshia001** [comment]
@jamiemitchellconsultants why not... Don't expect too much though, I didn't really put that much thought into the code; I've been hoping for an official release of indexing ever since.

**Comment by jamiemitchellconsultants** [comment]
It is a bit of a hatchet job...cutting out application logic... but my index grains are here https://github.com/jamiemitchellconsultants/Index

IndexInterface and IndexImplementation do the work.

Generally for a grain that implements a domain object.... e.g. Customer
I have CustomerGrain, CustomerIndex and CustomerManager

There is quite a bit of boilerplate ...mostly generated by T4 templates (not included)

My next step is to serialise Linq so that a client can send a predicate to the grain

**Comment by philbe** [comment]
@KSemenenko @Arshia001 @jamiemitchellconsultants See my comment [here](https://github.com/dotnet/orleans/issues/5281#issuecomment-450514056)

**Comment by Arshia001** [comment]
I know it's been 2 days and ~10 hours, I was having laptop problems... anyway, [here's mine](https://github.com/Arshia001/Indexing). There's unique and non-unique, and there's O(1) buckets.

**Comment by Arshia001** [comment]
Also, @philbe, you mention writing tests for the current code in your comment... I would be interested in doing that if time allows, given a few pointers on where to start and what to do.

**Comment by KSemenenko** [comment]
@Arshia001 @jamiemitchellconsultants let's help?
https://github.com/dotnet/orleans/issues/3413#issuecomment-450696037

**Comment by Arshia001** [comment]
With pleasure.
